agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpClientTracer.java

```diff
@@ -0,0 +1,179 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.auto.bootstrap.instrumentation.decorator;
+
+import static io.opentelemetry.context.ContextUtils.withScopedContext;
+import static io.opentelemetry.trace.TracingContextUtils.withSpan;
+
+import io.grpc.Context;
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.auto.config.Config;
+import io.opentelemetry.auto.instrumentation.api.MoreAttributes;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.trace.DefaultSpan;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.TracingContextUtils;
+import io.opentelemetry.trace.attributes.SemanticAttributes;
+import java.net.URI;
+import java.net.URISyntaxException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class HttpClientTracer<REQUEST, RESPONSE> extends BaseTracer {
+
+  private static final Logger log = LoggerFactory.getLogger(HttpClientTracer.class);
+
+  public static final String DEFAULT_SPAN_NAME = "HTTP request";
+
+  protected static final String USER_AGENT = "User-Agent";
+
+  protected abstract String method(REQUEST request);
+
+  protected abstract URI url(REQUEST request) throws URISyntaxException;
+
+  protected abstract Integer status(RESPONSE response);
+
+  protected abstract String requestHeader(REQUEST request, String name);
+
+  protected abstract String responseHeader(RESPONSE response, String name);
+
+  protected HttpTextFormat.Setter<REQUEST> getSetter() {
+    throw new UnsupportedOperationException();
+  }
+
+  public Span startSpan(REQUEST request) {
+    return startSpan(request, spanNameForRequest(request));
+  }
+
+  public Scope startScope(Span span, REQUEST request) {
+    Context context = withSpan(span, Context.current());
+    OpenTelemetry.getPropagators().getHttpTextFormat().inject(context, request, getSetter());
+    return withScopedContext(context);
+  }
+
+  // TODO should end methods remove SPAN attribute from request as well?
+  public void end(Span span, RESPONSE response) {
+    onResponse(span, response);
+    super.end(span);
+  }
+
+  public void endExceptionally(Span span, RESPONSE response, Throwable throwable) {
+    onResponse(span, response);
+    super.endExceptionally(span, throwable);
+  }
+
+  /**
+   * Returns a new client {@link Span} if there is no client {@link Span} in the current {@link
+   * Context}, or an invalid {@link Span} otherwise.
+   */
+  private Span startSpan(REQUEST request, String name) {
+    Context context = Context.current();
+    Span clientSpan = ClientDecorator.CONTEXT_CLIENT_SPAN_KEY.get(context);
```

Comment:
@trask In case of SERVER spans you wanted to let instrumentations to check for a presence of SERVER span and for them to decide to skip the call to `startSpan`. Here it is left for the Tracer to decide. We have to choose one consistent pattern.

I prefer to encapsulate this decision, as it is made here.

Reply:
ok, i think you and @anuraaga (https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/904#issuecomment-669684965) have convinced me